File: 19.remove-nth-node-from-end-of-list.js
// @leet start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
	let dummy = new ListNode(0);
	dummy.next = head;
	let fast = dummy;
	let slow = dummy;
	for(let i=0; i<=n; i++){
		fast = fast.next;
	}
	while(fast){
		fast = fast.next;
		slow = slow.next;
	}
	slow.next = slow.next.next;
	return dummy.next;
};
var reverse= function(head){
	let prev = null;
	let current = head;
	while(current){
		let next = current.next;
		current.next = prev;
		prev = current;
		current = next;
	}
	return prev;
}

// } @leet end

File: 19.remove-nth-node-from-end-of-list.c

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *removeNthFromEnd(struct ListNode *head, int n) {
    struct ListNode *slow = head;
    struct ListNode *fast = head;
    struct ListNode *prev = NULL;
    for (int i = 0; i < n; i++) {
        fast = fast->next;
    }
    while (fast) {
        prev = slow;
        slow = slow->next;
        fast = fast->next;
    }
    if (prev) {
        prev->next = slow->next;
    } else {
        head = head->next;
    }
    return head;
}
